planner/core/find_best_task.go

```diff
@@ -407,17 +420,163 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
 			}
 			continue
 		}
-		idxTask, err := ds.convertToIndexScan(prop, candidate)
+		if !isIdxMergePath {
+			idxTask, err := ds.convertToIndexScan(prop, candidate)
+			if err != nil {
+				return nil, err
+			}
+			if idxTask.cost() < t.cost() {
+				t = idxTask
+			}
+			continue
+		}
+		idxMergeTask, err := ds.convertToIndexMergeScan(prop, candidate)
 		if err != nil {
 			return nil, err
 		}
-		if idxTask.cost() < t.cost() {
-			t = idxTask
+		if idxMergeTask.cost() < t.cost() {
+			t = idxMergeTask
 		}
 	}
 	return
 }
 
+func (ds *DataSource) convertToIndexMergeScan(prop *property.PhysicalProperty, candidate *candidatePath) (task task, err error) {
+	if prop.TaskTp == property.CopSingleReadTaskType {
+		return invalidTask, nil
+	}
+	if !prop.IsEmpty() && !candidate.isMatchProp {
+		return invalidTask, nil
+	}
+	path := candidate.path
+	totalRowCount := 0.0
+	partPaths := make([]PhysicalPlan, 0, len(path.partialIndexPaths))
+	cop := &copTask{}
+	for _, partPath := range path.partialIndexPaths {
+		if partPath.isTablePath {
+			ts := PhysicalTableScan{
+				Table:           ds.tableInfo,
+				Columns:         ds.Columns,
+				TableAsName:     ds.TableAsName,
+				DBName:          ds.DBName,
+				isPartition:     ds.isPartition,
+				physicalTableID: ds.physicalTableID,
+			}.Init(ds.ctx)
+			ts.SetSchema(ds.schema)
+			if ts.Table.PKIsHandle {
+				if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+					if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+						ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+					}
+				}
+			}
+			path := candidate.path
+			ts.Ranges = path.ranges
+			ts.AccessCondition, ts.filterCondition = path.accessConds, path.tableFilters
+			rowCount := path.countAfterAccess
```

Comment:
It is pretty tricky to decide the row count of this table scan, we cannot use `path.countAfterAccess`, since we don't set it in `buildIndexMergeOrPath` at all. We cannot use `ds.stats.RowCount` neither, because `path.tableFilters` may be not empty. Currently, for DNF filters like `a < 1 or b < 1`, we are assuming column  a and column b are independent, and sum row_count(a<1) and row_count(b<1) as row_count(a<1 or b<1), to make it consistent, we can sum up  `countAfterAccess` of all partPaths as the row count of table scan. Further more, we need to consider `prop.ExpectedCnt` as well and multiply a selectivity factor if it is smaller than `ds.stats.RowCount`.

Reply:
https://github.com/pingcap/tidb/blob/85de5df03fd47f80b90c36a63b7d213aeda4468b/planner/core/stats.go#L178
We call this function to generate possible partial path. It can be seen that we have set related variables.